File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<Character>();
        for (int i =0 ;i<s.length();i++)
        {
            char m = s.charAt(i);
            if (m == '(' || m == '{' || m == '[')
            {
                st.push(m);
            }
            else
            {
                if (st.isEmpty() == true)
                {
                    return false;
                }
                else if (isMatching(st.peek(), m) == false)
                {
                    return false;
                }
                else
                {
                    st.pop();
                }
            }   
        }
        if (st.isEmpty() == true)
        {
            return true;
        }
        return false;
    }
    public boolean isMatching(Character a, Character b)
    {
        return ((a == '(' && b == ')') || (a == '{' && b == '}') || (a == '[' && b == ']'));
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        hashmap = {')': '(', '}': '{', ']': '['}
        for schar in s:
            if len(stack) == 0:
                stack.append(schar)
            else:
                if schar in hashmap:
                    if stack[-1] == hashmap[schar]:
                        stack.pop()
                    else:
                        return False
                else:
                    stack.append(schar)
        return len(stack) == 0